Sources/NIO/CircularBuffer.swift

```diff
@@ -50,36 +50,63 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
     ///
     /// Amortized *O(1)*
     public mutating func append(_ value: E) {
+        return self.addLast(value)
+    }
+
+    /// Add an element to the end of the ring buffer.
+    ///
+    /// Amortized *O(1)*
+    public mutating func addLast(_ value: E) {
         self.buffer[self.tailIdx] = value
         self.tailIdx = (self.tailIdx + 1) & self.mask
 
         if self.headIdx == self.tailIdx {
             // No more room left for another append so grow the buffer now.
-            var newBacking: ContiguousArray<E?> = []
-            let newCapacity = self.buffer.count << 1 // Double the storage.
-            precondition(newCapacity > 0, "Can't double capacity of \(self.buffer.count)")
-            assert(newCapacity % 2 == 0)
-
-            newBacking.reserveCapacity(newCapacity)
-            newBacking.append(contentsOf: self.buffer[self.headIdx..<self.buffer.count])
-            if self.headIdx > 0 {
-                newBacking.append(contentsOf: self.buffer[0..<self.headIdx])
-            }
-            newBacking.append(contentsOf: repeatElement(nil, count: newCapacity - newBacking.count))
-            self.tailIdx = self.buffer.count
-            self.headIdx = 0
-            self.buffer = newBacking
+            self.doubleCapacity()
+        }
+    }
+
+    /// Add an element to the front of the ring buffer.
+    ///
+    /// Amortized *O(1)*
+    public mutating func addFirst(_ value: E) {
```

Comment:
Swift doesn't really seem to use `add...`. I would suggest `prepend` maybe as it's the opposite of `append`? CC @lorentey, @moiseev, @airspeedswift

Reply:
also: what protocols should this auto expanding ring buffer implement?

Reply:
Appending to the front gets us to `BidirectionalCollection`. Now that we're at that, we can actually hop the next level up to `RandomAccessCollection`. Given the way this behaves, we could also give it `RangeReplaceableCollection` too, which is a pretty easy cousin of `RandomAccessCollection`.

So I'd recommend adding conformances to those two, though we may want to do them in a different PR.

Reply:
@Lukasa isn't `RangeReplaceableCollection` a thing where you can replace any range by any other range? So for example put 5 elements in a space where there used to be 3 ?

Like `thing[1...3] = ["foo", "bar", "buz", "qux", "quux"]` ?

Reply:
Yeah, and our buffer can do that. It's no less efficient at it than a regular array.

Reply:
sure but we have not implementation for that just yet so I'd do that whenever...

Reply:
Oh agreed, I think both of the new conformances should come in separate PRs to reduce scope. :smile:

Reply:
@weissi `prepend` sounds good to me as well... Waiting for @lorentey, @moiseev, @airspeedswift tho

Reply:
I vote for `prepend`/`append`.

Reply:
Thinking more about the right standard protocols for this type, looks like it should be possible to make `CircularBuffer` into a slice type. This way (after the `BidirectionalCollection` requirements are implemented) `removeLast`, `removeLast(_:)`, and `popLast()` default implementations will become automagically available.

Reply:
Ahh... No. You can't do that, because `CircularBuffer.startIndex` is always `0`. How attached are we to zero-based indexing for the `CircularBuffer`? ;-)

Reply:
I guess that (startIndex=0) wouldn't matter much. (though Redi/S lists also have that getrange-family of operations, but that could be mapped in the frontend).

The thing I'm wondering is whether there should also be a `prepend(contentsOf:)` which reserves the proper capacity.

Reply:
From what I can see, `CircularBuffer` does not have even an `append(contentsOf:)` right now.

I had this idea for a moment, but then I quickly hit a wall trying to name `prepend(contentsOf:)` correctly and specify the behavior when `tailIdx` reaches the `headIdx`, so I decided not to =)

Going back to indices. FWIW, in the collection types provided by the standard library there are no guarantees about the values of indices, so one should always use `startIndex`/`endIndex` combo and `index` family of methods, which might be annoying, but is the only guaranteed way of writing generic code over collections, even when `Index: FixedWidthInteger` or `Index == Int`.

Reply:
@moiseev Many thanks! I'm not personally attached to the 0 based indexing and I think we can change it as SemVer patch I think. What do @normanmaurer  & @Lukasa  think?

Reply:
and obvs I'm fine with prepend

Reply:
@weissi sounds good to me... That said I think we can pull in the PR as it is now... WDYT ?